### ISSUE TITLE
SPFx - How to load jsom dependencies in a proper way

### ISSUE THREAD

**Issue by emaborsa** [SharePoint/sp-dev-docs#8989]
### Target SharePoint environment

SharePoint Online

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Developer environment

Windows

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [ ] 💥 Microsoft Edge
- [ ] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [x] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

```
 "dependencies": {
    "@microsoft/sp-core-library": "1.17.1",
    "@microsoft/sp-loader": "^1.17.2",
    "@microsoft/sp-lodash-subset": "1.17.1",
    "@microsoft/sp-office-ui-fabric-core": "1.17.1",
    "@microsoft/sp-property-pane": "1.17.1",
    "@microsoft/sp-webpart-base": "1.17.1",
    "classnames": "^2.3.2",
    "office-ui-fabric-react": "^7.199.1",
    "papaparse": "^5.4.1",
    "react": "17.0.1",
    "react-dom": "17.0.1",
    "tslib": "2.3.1"
  },
  "devDependencies": {
    "@microsoft/eslint-config-spfx": "1.17.1",
    "@microsoft/eslint-plugin-spfx": "1.17.1",
    "@microsoft/rush-stack-compiler-4.5": "0.4.0",
    "@microsoft/sp-build-web": "1.17.1",
    "@microsoft/sp-module-interfaces": "1.17.1",
    "@rushstack/eslint-config": "2.5.1",
    "@types/microsoft-ajax": "^0.0.37",
    "@types/papaparse": "^5.3.7",
    "@types/react": "17.0.45",
    "@types/react-dom": "^17.0.17",
    "@types/sharepoint": "^2016.1.11",
    "@types/webpack-env": "~1.15.2",
    "ajv": "^6.12.5",
    "eslint": "8.7.0",
    "eslint-plugin-react-hooks": "4.3.0",
    "gulp": "4.0.2",
    "typescript": "4.5.5"
  }
```

### Describe the bug / error

I am getting crazy trying to load scripts like `Runtime.js` and `SP.js`. In the beginning I added the dependencies in the config.json as mentioned [here](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/guidance/connect-to-sharepoint-using-jsom).

I ended up to [load the SharePoint JSOM scripts with the SPComponentLoader](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/guidance/connect-to-sharepoint-using-jsom#load-the-sharepoint-jsom-scripts-with-the-spcomponentloader). In the debug it works, once deployed without **-p**, I get:

```
SP.Runtime.js:2 Uncaught Error: SP.Runtime.js is already loaded
Uncaught RangeError: Maximum call stack size exceeded
    at Boolean (<anonymous>)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180299)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
```

If I package the solution with **-p** I get:

```
GET https://relative-path.invalid/sp-init.js net::ERR_NAME_NOT_RESOLVED
GET https://relative-path.invalid/sp-runtime.js net::ERR_NAME_NOT_RESOLVED
GET https://relative-path.invalid/sharepoint.js net::ERR_NAME_NOT_RESOLVED
GET https://relative-path.invalid/microsoft-ajax.js net::ERR_NAME_NOT_RESOLVED
```

### Steps to reproduce

Actually, I am looking for the right way to load these dependencies.

### Expected behavior

Have the code running in all three scenarios.

**Comment by emaborsa** [comment]
I figured out, that using the flag `--build`, like `gulp build --debug` and `gulp package-solution --debug`, work.

**Comment by emaborsa** [comment]
As suggested by another user on another forum:

```

    if (window.location.pathname.indexOf("/_layouts/15/workbench.aspx") > -1) {
      SPComponentLoader.loadScript("/_layouts/15/init.js", {
        globalExportsName: "$_global_init",
      })
        .then((): Promise<{}> => {
          return SPComponentLoader.loadScript("/_layouts/15/MicrosoftAjax.js", {
            globalExportsName: "Sys",
          });
        })

        .then((): Promise<{}> => {
          return SPComponentLoader.loadScript("/_layouts/15/SP.Runtime.js", {
            globalExportsName: "SP",
          });
        })

        .then((): Promise<{}> => {
          return SPComponentLoader.loadScript("/_layouts/15/SP.js", {
            globalExportsName: "SP",
          });
        });
    } else {
      SPComponentLoader.loadScript("/_layouts/15/init.js", {
        globalExportsName: "$_global_init",
      });
    }
```

It works, really ugly... wonder if Microsoft has a clean solution.

**Comment by AJIXuMuK** [comment]
@emaborsa - do you have a solution you could share?
As mentioned in the docs you referenced you can use either declarative or code approach to reference the jsom libraries.
but not both.

**Comment by emaborsa** [comment]
Ok, lets start with the references in the `config.json` externals:

https://github.com/HGV/SPO.JSOM.Test

1. check out the repo
2. checkout using-externals
3. edit the `my.sharepoint.com` with your host
4. npm i
5. gulp serve
6. add the web part and click on `+ Click here`
7. I get the alert `ReferenceError: SP is not defined`

Now move using `ComponentLoader.loadScript`:

1. checkout using-component-loader
2. gulp serve
3. click on `+ Click here`
4. should alert `It works!`
5. gulp bundle -p
6. gulp package-solution -p
7. properly install solution and add web part
8. Add loading, before clicking ``+ Click here` I get the following error:

```
SP.Runtime.js:2 Uncaught Error: SP.Runtime.js is already loaded
(anonymous) @ SP.Runtime.js:2
init.js:1 Uncaught RangeError: Maximum call stack size exceeded
    at Boolean (<anonymous>)
    at ULSSendExceptionImpl (init.js:1:106252)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
```

**Comment by AJIXuMuK** [comment]
@emaborsa - I'm not seeing [this part](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/guidance/connect-to-sharepoint-using-jsom#reference-sharepoint-jsom-scripts-in-a-react-component) in your project:
```
require('sp-init');
require('microsoft-ajax');
require('sp-runtime');
require('sharepoint');
```

**Comment by emaborsa** [comment]
@AJIXuMuK ,
did you try to clone and fix it? 

1. `checkout using-externals`
2. add `
import "microsoft-ajax";
import "sharepoint";
import "sp-init";
import "sp-runtime";` in `Listview.tsx`
3. `gulp serve`
4. click on `+ Click here`
5. `It works!`
6. `gulp bundle -p`
7. `gulp package-solution -p`
8. properly install solution and add web part
9. At loading, before clicking `+ Click here` I get the following error:

```
SP.Runtime.js:2 Uncaught Error: SP.Runtime.js is already loaded
(anonymous) @ SP.Runtime.js:2
init.js:1 Uncaught RangeError: Maximum call stack size exceeded
    at Boolean (<anonymous>)
    at ULSSendExceptionImpl (init.js:1:106252)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
    at ULSSendExceptionImpl (init.js:1:106758)
    at ULSOnError (init.js:1:106829)
    at SupplementDefaultMethod.window.<computed> [as OriginalOnError] (init.js:1:180335)
```

**Comment by emaborsa** [comment]
@AJIXuMuK any update?

**Comment by AJIXuMuK** [comment]
@emaborsa - sorry for the really long delay.
I tested both branches in your sample with the steps you provided.

I'm not experiencing the issue for the `using-externals`
![image](https://github.com/SharePoint/sp-dev-docs/assets/17036219/cf757ebd-2137-4a6c-88b2-037cb1a3d5f8)

for the `using-component-loader` I needed to remove `import` statements as you don't have the modules loaded in that case.
After that I can both debug and deploy the solution without any issues.
